cpp/src/io/csv/csv_writer.cu

```diff
@@ -323,44 +327,65 @@ gdf_error write_csv(csv_write_arg* args)
         //
         // This is essentially an exclusive-scan (prefix-sum) across columns.
         // Moving left-to-right, add up each column and carry each value to the next column.
-        // Looks like we could transpose the matrix, scan it, and then untranspose it.
-        // Should be able to parallelize the math for this -- will look at prefix-sum algorithms.
+        // We do this in 3 kernel calls below.
         //
-        std::vector<char> buffer(memsize+1);
-        std::vector<size_t> string_locations(rows*count); // all the memory pointers for each column
+        rmm::device_vector<size_t> string_locations(rows*count); // all the memory pointers for each column
         string_locations[0] = 0; // first one is always 0
-        // compute offsets as described above into locations matrix
-        size_t offset = 0;
-        for( gdf_size_type jdx=0; jdx < rows; ++jdx )
+        size_t* d_string_locations = string_locations.data().get();
         {
-            // add up column values for each row
-            // this is essentially an exclusive-scan across columns
-            string_locations[jdx] = (size_t)(buffer.data() + offset); // initialize first item
+            rmm::device_vector<size_t> sums(rows);
+            size_t* d_sums = sums.data().get();
+            // do vertical scan -- add up lengths in each column
```

Comment:
Note that the memory access pattern here is not ideal but it's no worse than the striding done in the for_each_n calls, so overall it should be better. 

We could do even better by doing a transpose via shared memory first to eliminate all strides access. (https://devblogs.nvidia.com/efficient-matrix-transpose-cuda-cc/) However it would not be possible to fuse this with the first kernel inside the thrust scan (because it requires a `__syncthreads()`), so it would require writing a custom kernel.

Reply:
Thanks for reviewing this. Unfortunately, the single call the `transform_exclusive_scan` will not be enough. The transpose is needed only because the scan requires it. But the `d_string_locations` must now be _untransposed_ to get the result we need (example shown in the comments). The following change to your code snippet will work:
```
{
    rmm::device_vector<size_t> tmp_locations(rows*count);
    size_t* d_tmp_locations = tmp_locations.data().get();
    auto transpose_length = [d_string_lengths, rows, count] __device__ (size_t i) {
        size_t row = i % count;
        size_t col = i / count;
        size_t transposed_index = row * rows + col;
        return d_string_lengths[transposed_index]; 
    };
    thrust::transform_exclusive_scan(execpol->on(0),
                        thrust::make_counting_iterator<size_t>(0),
                        thrust::make_counting_iterator<size_t>(rows*count),
                        d_tmp_locations,
                        transpose_length, 0, thrust::plus<size_t>());
    auto untranspose_location = [d_tmp_locations, rows, count] __device__ (size_t i) {
        size_t row = i % rows;
        size_t col = i / rows;
        size_t transposed_index = row * count + col;
        return d_tmp_locations[transposed_index]; 
    };
    thrust::transform(execpol->on(0), 
                        thrust::make_counting_iterator<size_t>(0),
                        thrust::make_counting_iterator<size_t>(rows*count),
                        d_string_locations, untranspose_location );
}
```
Note that a temporary buffer `tmp_locations` is required to hold the intermediate transposed/scanned matrix which is then used for the extra `transform` method to transpose the result correctly into `d_string_locations`. The extra temporary buffer, therefore costs (rows*count) more memory. Also, I measured this with a simple test using the gutenberg dataset (~20 million rows with 4 columns) and found it to be about 60% slower the current code logic in this PR. 

Therefore, I'm inclined to add this a comment and not change the existing code logic. Also, this part of the code was about 10% of the pipeline (in the original 0.9-0.10 code base) but it is now only about 1% with other changes in this PR. Even major improvements to this part of the logic will not effect on the overall performance.

Reply:
Ah, good to know, thanks for experimenting. OK, well I have some simpler suggestions if this code is going to stick around.